client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/operations/RetryOnFailureOperation.java

```diff
@@ -165,16 +164,16 @@ protected Throwable handleException(Throwable cause, ChannelHandlerContext ctx,
          if (address != null) {
             updateFailedServers(address);
          }
-         if (ctx != null) {
+         if (channel != null) {
             // We need to remove decoder even if we're about to close the channel
             // because otherwise we would be notified through channelInactive and we would retry (again).
-            if (ctx.pipeline().get(HeaderDecoder.NAME) != null) {
-               ctx.pipeline().remove(HeaderDecoder.NAME);
+            if (channel.pipeline().get(HeaderDecoder.NAME) != null) {
```

Comment:
@rvansa there is a problem here: event listener re-registrations are triggered by HeaderDecoder's channelInactive, but removing the handler from the pipeline, causes the subsequent channel.close() to skip that

Reply:
Not sure if I full understand, but you could invoke the same login as in channelInactive in `handlerRemoved`

Reply:
Done